### PR TITLE
Grue changes: different moulting mechanics, exposure-stacking light damage, basic NPC AI, various other tweaks.

### PR DESCRIPTION
**Pull request by Hinaichigo**
In accordance with feedback from [this thread](http://ss13.moe/forum/viewtopic.php?f=4&t=6167), this refactors the grue light-based damage mechanics so that they take more damage per tick the longer they have recently been exposed to light, making brief sallies out into the light less punishing but continually increasing damage the longer the grue remains exposed.

Also adds some basic AI to NPC grues that prevents them from wandering into light, and lets them moult, eat, and reproduce.  The eggs will only hatch if they successfully recruit a player ghost though.

Also removes the shadow power mechanic and gives juveniles the ability to eat sentient beings as well, and successfully doing so makes them stronger and gives them the nutrients required to moult into adulthood. So they are more like weaker adults that can't lay eggs, and now aren't incentivized just to lurk in the dark until they are fully mature. Hopefully this makes them more interesting to play as.

Edit: I tagged this incorrectly
[content][tweak]

:cl:
 * rscadd: Grues now take initially lower but increasing damage the longer they spend exposed to light.
 * rscadd: Grues get a temporary healing boost after feeding.
 * rscadd: Juvenile grues can now eat like adults and must feed once in order to moult into an adult.  
 * rscadd: Added basic AI for NPC grues.
 * rscdel: Removed shadow power mechanic.
 * tweak: Grues heal faster in the dark.
 * tweak: Grues can pull other creatures and objects.
 * tweak: Grues can eat a little bit faster.

### CODE REVIEW

**Comment by Eneocho** [comment]
>The eggs will only hatch if they successfully recruit a player ghost though.

I was kinda looking forward to NPC gruening, like the NPC spidering that happens after a player spider lays eggs.

**Comment by Hinaichigo** [comment]
> > The eggs will only hatch if they successfully recruit a player ghost though.
> 
> I was kinda looking forward to NPC gruening, like the NPC spidering that happens after a player spider lays eggs.

I considered it as well but cycling dead players back into the round seems more fun, maybe.

**Comment by Eneocho** [comment]
make it spawn as simplemob if there was a mobcall but nobody joined.

**Comment by Eneocho** [comment]
Can you increase the door opening speed? I observed a grue and it took way too long for the guy to open doors as a new adult. 
Apparently it gets faster once you eat 5 players, but by then you don't need to open doors, you can crush walls.

And the waiting time was too long as well, it was straight up 10 minutes of camping on maintenance for the guy to accumulate enough power to go adult.

**Comment by WindowsErrors** [comment]
out of all the 10,000 human languages that eneocho could have spoken they decided to speak facts

if you think about it in terms of other reproducing maintenance-prowler antagonists like xenos or spiders, grues are definitely the slowest since they have the least benefits that come to them already being pre-established when you spawn in (like what, you get a deeper ass crack of maintenance to grow for 10 minutes in?) as opposed to xenos (the increased plasma output of the hive you spawn in means the only real waiting being done is by the original queen who set your place up and waited for you to sign up) or spiders (who are already going to spawn in areas spread out across the station, leaving very little downtime for you or the crew when more than one of them gets swole and in control)

**Comment by Hinaichigo** [comment]
> Can you increase the door opening speed? I observed a grue and it took way too long for the guy to open doors as a new adult. Apparently it gets faster once you eat 5 players, but by then you don't need to open doors, you can crush walls.

Currently it takes a juvenile 10 seconds and a yet-to-eat adult 5 seconds, removing 2 seconds every time the adult eats, so it becomes instant after 3 eatings. I think reducing it to 5 for juveniles and 3 for adults (-1 second each feeding) could be a good balance, so I went ahead and did that. I do want to avoid it being instant though so they have to be a bit more cautious about venturing into the light.

As for the waiting time, the juvenile state is sort of meant to be a hybrid state between the cautious, hiding mode of the larva and the more aggressive, hungry state of the adult. Not necessarily just for waiting around, though I guess people are incentivized to just hide and not make their presence known until they reach their most powerful form.
One solution to this I was thinking of was giving the juvenile the Eat ability as well and they would instantly morph into an adult upon eating a sentient, (or maybe be able to moult much faster than normal, or get maxed shadow power so they can moult immediately) so there would be some incentive for riskier plays and not just hiding to fully power up before being aggressive. Another option is to make moulting into an adult require the juvenile grue to eat a sentient first.

Some numbers for the time to reach adult:
In full darkness a grue gains 3 shadow power per second, so it takes:
~83 seconds (1:23) to get the 250 shadow power necessary to moult into a juvenile.
30 seconds to moult
 ~167 (2:47) seconds to get the 500 shadow power necessary to moult into an adult
30 seconds to moult.
So it takes 5 minutes and 10 seconds to go from larva to adult, assuming they were in full darkness the whole time.

Let me know what kind of adjustments you'd like to see on those times, or if there's some change that could make things more interesting for the not-yet-fully-mature grue.

**Comment by Eneocho** [comment]
You could add some minor buffs if the grue fed as a teen (better light endurance? eating lights for a sec?), so there's an incentive.
Prime Grues are those that fed as teens and have the extra stuff. Regular grues are those that just became adults by waiting.

At adult, the guy I observed couldn't heal the damage in time, it ticked too slow and since he couldn't eat a single sentient, he also didn't get any healing (maybe add healing via eating monkeys?). 
You could reuse the shadowpower from baby and teen as a ability "fuel". You actively accumulate shadowpower as adult too and can use it to power some more advanced shit, like the Umbra (former grue) darkness area power, enhanced healing, leaping (tackling but you don't get stunned), etc.

**Comment by Eneocho** [comment]
Also, I'd slice the waiting times for baby in half, the teen one by a little too.

**Comment by Hinaichigo** [comment]
Interesting ideas. One thing I almost added to this PR was a temporary healing boost after a successful feeding, but decided against it. It was such that after eating you would heal an extra 2 health per second for 25 seconds. What do you think about that? And what do you and others think about having there not be a sentient check for feeding in general. So if a grue breaks into virology or genetics during a power outage or something you have a catastrophe on your hands.

Overall I erred on the side of making them underpowered at first (the EI NATH thing notwithstanding), but I guess, indeed, people want them to be somewhat stronger?

> Also, I'd slice the waiting times for baby in half, the teen one by a little too.

The shadow power time, not the moulting time, right? I like having a little bit of a substantial moulting time so that it forces them to be vulnerable for a bit. It also seems like it can have some interesting dynamics with herds of grues, active ones protecting the moulting ones, etc.

**Comment by Eneocho** [comment]
The removal of the sentient check would be interesting, it'd become an alternate spider (since spiders just need to feed on living beings, no matter the sentience). 
I'd make the feeding on non-sentients contribute less than feeding on sentients. Say, 2 monkeys = 1 sentient for point purposes, so you CAN get to the advanced levels, but you are encouraged to fight against crew for faster growth.
That way it'd be a viable if slow way to reach fuckgrue levels.
Do add the healingbuff, either via spell and accumulated points or via feeding on anything.

>The shadow power time, not the moulting time, right?

Yes, the shadow time.

**Comment by Hinaichigo** [comment]
What do you think about increasing brute resistance as grues eat more?

**Comment by Eneocho** [comment]
Not sure. You mean as they eat more and more? You could increase the resistance by just a bit whenever they feed, so the 5 sentients fuckgrue is a force to be reckoned with and not a casual fight (similar to queens in both beno and spiders).
Try it, if it doesn't work, you can just remove it later.

**Comment by Hinaichigo** [comment]
Doubled the base healing rate and added 75 health worth of healing over 25 seconds per sentient feed. Also reduced the larva->juvenile shadow power-time from 83 to 50 seconds and juvenile to 100 seconds (1:40).
I hesitate to make it too fast though, otherwise people might not try to do anything as larva and juveniles and just wait until they can moult, type of thing.

**Comment by WindowsErrors** [comment]
but isn't that what the point of a larva is? laying low until you're ready to strike out?

**Comment by Hinaichigo** [comment]
Good point. Also the ventcrawling lets you scout the map and see where you might want to moult and base your operations in. I'll set it to 30 seconds for larvae and 1:30 for juveniles.

**Comment by Hinaichigo** [comment]
Changed it so that juvenile grues can also Eat sentients. It takes them ~2:30 to get enough shadowpower to moult but eating a sentient will give half of the necessary shadowpower over 25 seconds. So this should hopefully make juveniles more like weaker, non-egglaying adults and maybe more dynamic to play as.
Also set the airlock forcing time to instant for adults and 3 seconds for juveniles (becomes instant if they eat someone).

**Comment by Hinaichigo** [comment]
Okay, I changed the shadowpower mechanic into "nutritive energy." Basically a larva starts out already nourished from the egg yolk so they can moult whenever they want. It's up to the player if they want to moult where they spawn or if they want to scout the map for a more optimal location by ventcrawling and exploring.
And juveniles start out hungry and can only moult after consuming one sentient being. They can't reproduce until doing so either. Now when the grue herd kills someone they might end up squabbling over who gets to eat the carcass, which could be interesting and adds to the hunger vibes; a juvenile could eat it to grow into an adult, or an adult could eat it to get even stronger.

**Comment by Eneocho** [comment]
Very nice.
Here's hoping grue get some special powers eventually, but that's for another PR.

File: code/modules/mob/living/simple_animal/grue_egg.dm [hunk 1]
```diff
@@ -35,6 +35,11 @@
 	var/last_ping_time = 0
 	var/ping_cooldown = 50
 
+	var/accum_light_expos_mult= 1 //used to scale light damage the longer the grue egg is exposed to light
+	var/list/accum_light_expos_gain_dark_dim_light=list(-3,-1,1) //light damage rate increases the longer the grue egg is exposed to light, but this effect dissipates after going back into darkness
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/grue_egg.dm]: This var and others are duplicated between `grue_egg` and `grue`.
Have you considered using inheritance to avoid repeating yourself?

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 2]
```diff
@@ -137,13 +166,81 @@
 	regular_hud_updates()
 	standard_damage_overlay_updates()
 
+
+	if((stat!=DEAD) && (stat!=UNCONSCIOUS) && !busy && !ismoulting && !client && !mind && ckey==null) //Checks for AI
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
	if((stat == CONSCIOUS) && !busy && !ismoulting && !client && !mind && !ckey) //Checks for AI
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 3]
```diff
@@ -137,13 +166,81 @@
...
+								potential_dests+=potential_dest
+					var/ourdest
+					if(respite_dests.len)
+						ourdest=(pick(respite_dests))
+					else if(potential_dests.len)
+						ourdest=(pick(potential_dests))
+					if(ourdest)
+						wander_move(ourdest)
+						turns_since_move = 0
+						INVOKE_EVENT(src, /event/after_move)
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]: `wander_move` will call `Move` which should invoke this event. Is this necessary for some reason?

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 4]
```diff
@@ -333,6 +439,27 @@
...
+/mob/living/simple_animal/hostile/grue/proc/get_light_damage()
+	return (current_brightness-bright_limit_drain) * accum_light_expos_mult * hd_mult * lightresist * (maxHealth/250) 	//scale light damage by: how bright the light is, amount of recent light exposure, the base multiplier, lifestage-dependent resistance, and normalize by max health,
+
+
+/mob/living/simple_animal/hostile/grue/proc/get_dark_heal()
+	return -1 * (bright_limit_gain-current_brightness) * hg_mult * regenbonus * (maxHealth/250)							//scale dark healing by: how dark it is, the base multiplier, the bonus based on sentient beings eaten, and normalize by max health.
+
+/mob/living/simple_animal/hostile/grue/proc/nutri_adjust()
+	switch(dark_dim_light)
+		if(0) //if dark
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]: These should use the newly-#defined macros

File: code/modules/mob/living/simple_animal/grue_egg.dm (hunk 1 above)
Comment (Hinaichigo) [reply code/modules/mob/living/simple_animal/grue_egg.dm]: One issue is that grue eggs are  `mob/living/simple_animal` while grues are `mob/living/simple_animal/hostile`, so I'm not sure what the ideal solution is.

File: code/modules/mob/living/simple_animal/hostile/grue.dm (hunk 3 above)
Comment (Hinaichigo) [reply code/modules/mob/living/simple_animal/hostile/grue.dm]: I copied it along from elsewhere, I'll remove it.

File: code/modules/mob/living/simple_animal/grue_egg.dm (hunk 1 above)
Comment (DamianX) [reply code/modules/mob/living/simple_animal/grue_egg.dm]: True. You could move these vars and procs to a datum and use it in both places.

Comment (Hinaichigo) [reply code/modules/mob/living/simple_animal/grue_egg.dm]: I think I got it set up as a datum.

**Comment by Hinaichigo** [comment]
Moved the light-related parameters to a datum shared by both grues and grue eggs.

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 5]
```diff
@@ -21,45 +21,112 @@
...
 	var/eatencount=0												//number of sentient carbons eaten, makes the grue more powerful
 	var/eatencharge=0												//power charged by eating sentient carbons, increments with eatencount but is spent on upgrades
 	var/spawncount=0												//how many eggs laid by this grue have successfully hatched
-	var/dark_dim_light=0 //darkness level currently the grue is currently exposed to, 0=nice and dark, 1=passably dim, 2=too bright
-	var/busy=0 //busy laying an egg
+
+	var/busy=0 //busy attempting to lay an egg or eat
 
 	var/eattime= 5 SECONDS //how long it takes to eat someone
+	var/digest = 0 //how many seconds of healing left after feeding
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]: This seems to be misleading. The value is not expressed in seconds but in number of Life() ticks, which currently happen every 2 seconds.

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 6]
```diff
@@ -21,45 +21,112 @@
...
+	var/digest_heal = -3 //how much health restored per second after feeding (negative heals)
+	var/digest_sp = 4 //how much nutritive energy gained per second after feeding
+
+	var/datum/grue_calc/grue/lightparams = new /datum/grue_calc/grue //used for light-related calculations
+
+	//AI related:
+	stop_automated_movement = 1 //has custom light-related wander movement
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
	stop_automated_movement = TRUE //has custom light-related wander movement
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 7]
```diff
@@ -21,45 +21,112 @@
...
+	wander=0
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
	wander = FALSE
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 8]
```diff
@@ -137,13 +210,80 @@
...
+
+	//Movement
+	//
+	//Wandering behavior
+	if((!client||deny_client_move) && !busy && !ismoulting && !anchored && (ckey == null) && !(flags & INVULNERABLE))
+		if(canmove)
+			turns_since_move++
+			if(turns_since_move >= turns_per_move)
+				if(!(stop_automated_movement_when_pulled && pulledby))
+					INVOKE_EVENT(src, /event/before_move)
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]: `wander_move` should call `Move` which should invoke this event, why is this here?

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 9]
```diff
@@ -137,13 +210,80 @@
...
+					var/list/potential_dests=list() //any potential wander destinations
+					var/list/respite_dests=list() //any potential wander destinations leading out of the light into dark/dim
+					for(var/thiscardinal in 1 to cardinal.len)
+						var/potential_dest = get_step(src, cardinal[thiscardinal])
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
					for(var/direction in cardinal)
						var/potential_dest = get_step(src, direction)
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 10]
```diff
@@ -137,13 +210,80 @@
...
+							if(get_ddl(potential_dest)<GRUE_LIGHT)
+								respite_dests+=potential_dest
+							else
+								potential_dests+=potential_dest
+						else //never wander from a dark/dim tile into light
+							if(get_ddl(potential_dest)<GRUE_LIGHT)
+								potential_dests+=potential_dest
+					var/ourdest
+					if(respite_dests.len)
+						ourdest=(pick(respite_dests))
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
						ourdest=pick(respite_dests)
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 11]
```diff
@@ -137,13 +210,80 @@
...
+					else if(potential_dests.len)
+						ourdest=(pick(potential_dests))
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
						ourdest=pick(potential_dests)
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 12]
```diff
@@ -21,45 +21,112 @@
...
+	var/base_melee_dam_lw = 3				//base melee damage lower
 
 	var/lifestage=GRUE_ADULT												 //1=baby grue, 2=grueling, 3=(mature) grue
 	var/eatencount=0												//number of sentient carbons eaten, makes the grue more powerful
 	var/eatencharge=0												//power charged by eating sentient carbons, increments with eatencount but is spent on upgrades
 	var/spawncount=0												//how many eggs laid by this grue have successfully hatched
-	var/dark_dim_light=0 //darkness level currently the grue is currently exposed to, 0=nice and dark, 1=passably dim, 2=too bright
-	var/busy=0 //busy laying an egg
+
+	var/busy=0 //busy attempting to lay an egg or eat
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
	var/busy=FALSE //busy attempting to lay an egg or eat
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 13]
```diff
@@ -410,50 +556,63 @@
...
 /mob/living/simple_animal/hostile/grue/proc/handle_feed(var/mob/living/E)
-	to_chat(src, "<span class='danger'>You open your mouth wide, preparing to eat [E]!</span>")
+	visible_message("<span class='danger'>\The [src] opens its mouth wide...</span>","<span class='danger'>You open your mouth wide, preparing to eat \the [E]!</span>")
 	busy=1
 	if(do_mob(src , E, eattime, eattime, 0)) //check on every tick
-		to_chat(src, "<span class='danger'>You have eaten [E]!</span>")
+		to_chat(src, "<span class='danger'>You have eaten \the [E]!</span>")
 		to_chat(E, "<span class='danger'>You have been eaten by a grue.</span>")
 
+		digest+=25 //add 25 seconds of increased healing + nutrition gain
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]: See earlier comment about this var. This will add 50 seconds, not 25, I think.

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 14]
```diff
@@ -410,50 +556,63 @@
...
 
-			regenbonus=regenbonus*1.5 //increased health regen in darkness
+	//speed bonus in dark and dim conditions
+	lightparams.speed_m_dark_dim_light[1]=max(1/2,lightparams.base_speed_m_dark_dim_light[1]/(1.2 ** eatencount))//faster in darkness
+	lightparams.speed_m_dark_dim_light[2]=max(5/8,lightparams.base_speed_m_dark_dim_light[2]/(1.2 ** eatencount))//faster in dim light
+
+
+	if(lifestage==GRUE_ADULT)
+		if(eatencount>=GRUE_WALLBREAK)
+			environment_smash_flags = environment_smash_flags | SMASH_WALLS
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
			environment_smash_flags |= SMASH_WALLS
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm [hunk 15]
```diff
@@ -410,50 +556,63 @@
...
+		if(eatencount>=GRUE_RWALLBREAK)
+			environment_smash_flags = environment_smash_flags | SMASH_RWALLS
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/grue.dm]:
```suggestion
			environment_smash_flags |= SMASH_RWALLS
```

File: code/modules/mob/living/simple_animal/hostile/grue.dm (hunk 13 above)
Comment (Hinaichigo) [reply code/modules/mob/living/simple_animal/hostile/grue.dm]: Oh, I had assumed Life() was 1 second.

**Comment by Eneocho** [comment]
Please make it so teen and adult grues can drag their victims. All other "monster" antags can.

**Comment by Hinaichigo** [comment]
> Please make it so teen and adult grues can drag their victims. All other "monster" antags can.

I'll take a look at doing that.

**Comment by Eneocho** [comment]
Could you eventually add that light sucking skill that humanoid grues (now umbra) had, so you can make a mad rush against someone with lights on them?

**Comment by Hinaichigo** [comment]
> Could you eventually add that light sucking skill that humanoid grues (now umbra) had, so you can make a mad rush against someone with lights on them?

I've been thinking about it and whether and how to implement that... though if I can get #32007 working it might make it less necessary, especially if grues can target specific bodyparts. Then they could bite someone's arm to make them drop their lantern, then smash the lantern once it's on the ground. 
Also, though it might be interesting mechanically, I hesitate because I wonder if the dark aura would be flavor-congrueous with the current iteration of grues; it would sort of make them more shadow-elementalish, which is something I've been avoiding.

**Comment by Eneocho** [comment]
I'd like to see it though, it would work mechanically and as an homage to the original grue (Umbra).
Try it, worst that can happen is it doesn't work and we revert it.

**Comment by Hinaichigo** [comment]
Maybe we can come up with a bunch of different powers post-larval grues can use, and the shadow aura could be one of them. If we add that we could also add invisibility in full (not dim) darkness. But like you said it's for a different PR.
I added the pulling.
Also, should umbras be renamed to gruekin or something like that, do you think? Gruoids..?

**Comment by Eneocho** [comment]
I think umbra is fine, though I use both for now since we're still at that stage where the two variants can be confused.

**Comment by Hinaichigo** [comment]
> I think umbra is fine, though I use both for now since we're still at that stage where the two variants can be confused.

By the way if you want to, please take a look at #31969.

**Review by DamianX — APPROVED** [review]
Is this ready?

**Comment by Hinaichigo** [comment]
> Is this ready?

Yes, ready.
